source/common/http/http1/codec_impl.cc

```diff
@@ -730,6 +730,29 @@ Envoy::StatusOr<int> ConnectionImpl::onHeadersCompleteBase() {
     handling_upgrade_ = true;
   }
 
+  // https://tools.ietf.org/html/rfc7230#section-3.3.3
+  // If a message is received with both a Transfer-Encoding and a
+  // Content-Length header field, the Transfer-Encoding overrides the
+  // Content-Length. Such a message might indicate an attempt to
+  // perform request smuggling (Section 9.5) or response splitting
+  // (Section 9.4) and ought to be handled as an error. A sender MUST
+  // remove the received Content-Length field prior to forwarding such
+  // a message.
+
+  // Reject message with Http::Code::BadRequest if both Transfer-Encoding and Content-Length
+  // headers are present or if allowed by http1 codec settings and 'Transfer-Encoding'
+  // is chunked - remove Content-Length and serve request.
+  if (parser_.uses_transfer_encoding != 0 && request_or_response_headers.ContentLength()) {
+    if ((parser_.flags & F_CHUNKED) && codec_settings_.allow_chunked_length_) {
+      request_or_response_headers.removeContentLength();
+    } else {
+      error_code_ = Http::Code::BadRequest;
```

Comment:
Do we have an integration test that covers this?    If not I'd like to make sure we add one.

Reply:
Added server/client integration tests for enabled allow_chunked_length_.
Existing smuggling test covers disabled allow_chunked_length_ logic.